janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java

```diff
@@ -416,17 +453,107 @@ public void bulkRequest(List<ElasticSearchMutation> requests, String ingestPipel
             APPEND_OP.apply(builder).append("refresh=").append(bulkRefresh);
         }
         builder.insert(0, REQUEST_SEPARATOR + "_bulk");
```

Comment:
Not related to your PR, but it seems it makes sense to move this line up (to be the next line after `StringBuilder` is created) to avoid array shifting operation. I.e. Just:
```
final StringBuilder builder = new StringBuilder();
builder.append(REQUEST_SEPARATOR).append("_bulk");
```

Reply:
Looked into this this for my follow-up, but it seems like this would change behavior?

https://github.com/JanusGraph/janusgraph/blob/d255396dcddf5c19317a873ff1ce16e31f77a304/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java#L105

`APPEND_OP` either prefixes a `?` or a `&` if the builder is empty. So as currently written I believe it's producing paths like `/_bulk?pipeline=foo&refresh=true`.

If we put `builder.append(REQUEST_SEPARATOR).append("_bulk");` further up, I think it'd produce paths like `/_bulk&pipeline=foo&refresh=true` which wouldn't be properly formatting the query parameters anymore.

How does this look if we just use the string builder for the query parameters, and then just directly build the path? I originally used another StringBuilder for the path variable, but IntelliJ's linter suggested just doing String concatenations.

```
final StringBuilder bulkRequestQueryParameters = new StringBuilder();
        if (ingestPipeline != null) {
            APPEND_OP.apply(bulkRequestQueryParameters).append("pipeline=").append(ingestPipeline);
        }
        if (bulkRefreshEnabled) {
            APPEND_OP.apply(bulkRequestQueryParameters).append("refresh=").append(bulkRefresh);
        }
        final String bulkRequestPath = REQUEST_SEPARATOR + "_bulk" + bulkRequestQueryParameters;
        return Pair.with(bulkRequestPath, outputStream.toByteArray());
```

Reply:
Oh. I missed that! I'm OK with leaving it as it is right now or updating it to the way you are suggesting. Both ways are OK by me.
btw. Usually `final` String concatenation is directly compiled to reuse `StringBuilder` by all modern Java compilers, so I believe that's why IntelliJ's suggested just using String concatenation here.
Anyways, the shift is probably too small to actually produce any meaningful performance differences, so I'm OK with either way as you prefer.